fl4health/losses/contrastive_loss.py

```diff
@@ -0,0 +1,61 @@
+import torch
+
+
+class ContrastiveLoss(torch.nn.Module):
+    def __init__(
+        self,
+        device: torch.device,
+        temperature: float = 0.5,
+    ) -> None:
+        """
+        This constrastive loss is implemented based on https://github.com/QinbinLi/MOON.
+        Contrastive loss aims to enhance the similarity between the features and their positive pairs
+        while reducing the similarity between the features and their negative pairs.
+
+        Args:
+            device (torch.device): device to use for computation
+            temperature (float): temperature to scale the logits
+        """
+
+        super().__init__()
+        self.device = device
+        self.temperature = temperature
+        self.cos_sim = torch.nn.CosineSimilarity(dim=-1).to(self.device)
+        self.ce_criterion = torch.nn.CrossEntropyLoss().to(self.device)
+
+    def forward(
+        self, features: torch.Tensor, positive_pairs: torch.Tensor, negative_pairs: torch.Tensor
+    ) -> torch.Tensor:
+        """
+        Compute the contrastive loss based on the features, positive pair and negative pairs. While every feature
+        has a positive pair, it can have multiple negative pairs. The loss is computed based on the similarity
+        between the feature and its positive pair.
+
+        Args:
+            features (torch.Tensor): Main features, shape (n_samples, n_features)
+            positive_pairs (torch.Tensor): Positive pair of main features, shape (1, n_samples, n_features)
+            negative_pairs (torch.Tensor): Negative pairs of main features, shape (n_pairs ,n_samples, n_features)
+
+        Returns:
+            torch.Tensor: Contrastive loss value
+        """
+        # TODO: We can extend it to support multiple positive pairs using multi-label classification
+
+        assert self.temperature is not None
+        if len(positive_pairs) != 1:
+            AssertionError(
+                """Each feature can have one positive pair.
```

Comment:
A few minor points:

1. I think, though not 100% certain, you need to use `raise AssertionError(...)` for this to actually throw?
2. I believe the triple quotes here will mean that any \n values you have will also be logged. So I think it's better to just do 
```
AssertionError(
    "Each feature can have one positive pair. "
    "Thus positive pairs should be a tensor of shape (1, n_samples, n_features) "
    f"rather than {positive_pairs.shape}"
)
```
3. It would likely be useful to the user to also show the shape of the tensor they entered. See above

Reply:
I agree with all of them. Thanks for recommendations